<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Module%201/images/IDSNlogo.png" width="300" alt="cognitiveclass.ai logo">
</center>

# Introduction  Notebook

Estimated time needed: **10** minutes

## Objectives

After completing this lab you will be able to:

*   Acquire data in various ways
*   Obtain insights from data with Pandas library


<h2>Table of Contents</h2>

<div class="alert alert-block alert-info" style="margin-top: 20px">
<ol>
    <li><a href="https://#data_acquisition">Data Acquisition</a>
    <li><a href="https://#basic_insight">Basic Insight of Dataset</a></li>
</ol>

</div>
<hr>


<h1 id="data_acquisition">Data Acquisition</h1>
<p>
There are various formats for a dataset: .csv, .json, .xlsx  etc. The dataset can be stored in different places, on your local machine or sometimes online.<br>

In this section, you will learn how to load a dataset into our Jupyter Notebook.<br>

In our case, the Automobile Dataset is an online source, and it is in a CSV (comma separated value) format. Let's use this dataset as an example to practice data reading.

<ul>
    <li>Data source: <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/APEBUSD_trades_1m.csv" target="_blank">https://1824251045.rsc.cdn77.org/web/algohouse/data/APEBUSD_trades_1m.csv</a></li>
    <li>Data type: csv</li>
</ul>
The Pandas Library is a useful tool that enables us to read various datasets into a dataframe; our Jupyter notebook platforms have a built-in <b>Pandas Library</b> so that all we need to do is import Pandas without installing.
</p>


you are running the lab in your  browser, so we will install the libraries using ``piplite``


In [4]:
#you are running the lab in your  browser, so we will install the libraries using ``piplite``
import piplite
import micropip
await piplite.install(['pandas'])
await piplite.install(['matplotlib'])
await piplite.install(['scipy'])
await piplite.install(['seaborn'])
await micropip.install(['ipywidgets'],keep_going=True)
await micropip.install(['tqdm'],keep_going=True)

If you run the lab locally using Anaconda, you can load the correct library and versions by uncommenting the following:


In [ ]:
#install specific version of libraries used in  lab
#! mamba install pandas==1.3.3  -y
#! mamba install numpy=1.21.2 -y

In [1]:
# import pandas library
import pandas as pd
import numpy as np

This function will download the dataset into your browser 


In [16]:
#This function will download the dataset into your browser 

from pyodide.http import pyfetch

async def download(url, filename):
    response = await pyfetch(url)
    if response.status == 200:
        with open(filename, "wb") as f:
            f.write(await response.bytes())

<h2>Read Data</h2>
<p>
We use <code>pandas.read_csv()</code> function to read the csv file. In the brackets, we put the file path along with a quotation mark so that pandas will read the file into a dataframe from that address. The file path can be either an URL or your local file address.<br>

Because the data does not include headers, we can add an argument <code>headers = None</code> inside the <code>read_csv()</code> method so that pandas will not automatically set the first row as a header.<br>

You can also assign the dataset to any variable you create.

</p>


In [4]:
path = "https://1824251045.rsc.cdn77.org/web/algohouse/data/APEBUSD_trades_1m.csv"

you will need to download the dataset; if you are running locally, please comment out the following 


In [2]:
#you will need to download the dataset; if you are running locally, please comment out the following 
await download(path, "APEBUSD_trades_1m.csv")
path = "APEBUSD_trades_1m.csv"

This dataset was hosted on IBM Cloud object. Click <a href="https://1824251045.rsc.cdn77.org/web/algohouse/data/index.html">HERE</a> for free storage.


In [3]:
# Import pandas library
import pandas as pd

# Read the online file by the URL provides above, and assign it to variable "df"

df = pd.read_csv(path, index_col=0)

After reading the dataset, we can use the <code>dataframe.head(n)</code> method to check the top n rows of the dataframe, where n is an integer. Contrary to <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> will show you the bottom n rows of the dataframe.


In [7]:
# # Added to replace actual columns with numbers, thus we can show how to assign columns 
# df.columns = range(len(df.columns))

In [4]:
# show the first 5 rows using dataframe.head() method
print("The first 5 rows of the dataframe") 
df.head(5)

The first 5 rows of the dataframe


ts   open   high    low  close   volume  rec_count  \
0  2022-11-11 14:38:00  3.096  3.096  3.076  3.077   2733.3         34   
1  2022-11-11 14:39:00  3.073  3.077  3.064  3.065   5353.1         40   
2  2022-11-11 14:40:00  3.068  3.075  3.062  3.075   5582.2         34   
3  2022-11-11 14:41:00  3.075  3.091  3.069  3.085  10358.0         43   
4  2022-11-11 14:42:00  3.085  3.093  3.074  3.074   9567.6         38   

   avg_price  
0   3.084732  
1   3.069644  
2   3.071029  
3   3.084393  
4   3.088095

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #1: </h1>
<b>Check the bottom 10 rows of data frame "df".</b>
</div>


In [10]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print("The last 10 rows of the dataframe\n")
df.tail(10)
```


<h3>Columnns</h3>
<p>
The columns are stored in <code>df.columns</code>
</p>
<p>
Let's capitalize every first letter
</p>


In [5]:
# create headers list
columns = list(df.columns)
columns = [col.capitalize() for col in columns]
print("Columns\n", columns)

Columns
 ['Ts', 'Open', 'High', 'Low', 'Close', 'Volume', 'Rec_count', 'Avg_price']


We replace headers and recheck our dataframe:


In [6]:
df.columns = columns
df.head(10)

Ts   Open   High    Low  Close   Volume  Rec_count  \
0  2022-11-11 14:38:00  3.096  3.096  3.076  3.077   2733.3         34   
1  2022-11-11 14:39:00  3.073  3.077  3.064  3.065   5353.1         40   
2  2022-11-11 14:40:00  3.068  3.075  3.062  3.075   5582.2         34   
3  2022-11-11 14:41:00  3.075  3.091  3.069  3.085  10358.0         43   
4  2022-11-11 14:42:00  3.085  3.093  3.074  3.074   9567.6         38   
5  2022-11-11 14:43:00  3.076  3.085  3.073  3.079   2212.5         27   
6  2022-11-11 14:44:00  3.082  3.094  3.082  3.092   8106.4         28   
7  2022-11-11 14:45:00  3.095  3.138  3.095  3.138  35545.1         67   
8  2022-11-11 14:46:00  3.137  3.154  3.120  3.153  17379.1         57   
9  2022-11-11 14:47:00  3.158  3.236  3.158  3.214  66603.7         75   

   Avg_price  
0   3.084732  
1   3.069644  
2   3.071029  
3   3.084393  
4   3.088095  
5   3.079011  
6   3.089105  
7   3.118701  
8   3.138483  
9   3.194934

<!-- We need to replace the "?" symbol with NaN so the dropna() can remove the missing values:
 -->

Now, we have successfully read the raw dataset and modified columns.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #2: </h1>
<b>Find the name of the columns of the dataframe.</b>
</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
print(df.columns)
```

</details>


<h2>Save Dataset</h2>
<p>
Correspondingly, Pandas enables us to save the dataset to csv. By using the <code>dataframe.to_csv()</code> method, you can add the file path and name along with quotation marks in the brackets.
</p>
<p>
For example, if you would save the dataframe <b>df</b> as <b>automobile.csv</b> to your local machine, you may use the syntax below, where <code>index = False</code> means the row names will not be written.
</p>


We can also read and save other file formats. We can use similar functions like **`pd.read_csv()`** and **`df.to_csv()`** for other data formats. The functions are listed in the following table:


<h2>Read/Save Other Data Formats</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Basic Insight of Dataset</h1>
<p>
After reading data into Pandas dataframe, it is time for us to explore the dataset.<br>

There are several ways to obtain essential insights of the data to help us better understand our dataset.

</p>


<h2>Data Types</h2>
<p>
Data has a variety of types.<br>

The main types stored in Pandas dataframes are <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> and <b>datetime64</b>. In order to better learn about each attribute, it is always good for us to know the data type of each column. In Pandas:

</p>


In [7]:
df.dtypes

Ts            object
Open         float64
High         float64
Low          float64
Close        float64
Volume       float64
Rec_count      int64
Avg_price    float64
dtype: object

A series with the data type of each column is returned.


<p>
As shown above, it is clear to see that the data type of "high" and "close" are <code>float64</code>, "rec_count" is <code>int_64</code>, and "ts" is <code>object</code>, what is not correct so we need to change this
</p>

In [8]:
df['Ts'] = pd.to_datetime(df['Ts'])

In [10]:
# check the data type of data frame "df" by .dtypes
print(df.dtypes)

Ts           datetime64[ns]
Open                float64
High                float64
Low                 float64
Close               float64
Volume              float64
Rec_count             int64
Avg_price           float64
dtype: object


<h2>Describe</h2>
If we would like to get a statistical summary of each column e.g. count, column mean value, column standard deviation, etc., we use the describe method:


This method will provide various summary statistics, excluding <code>NaN</code> (Not a Number) values.


In [11]:
df.describe()

Open          High           Low         Close         Volume  \
count  66528.000000  66528.000000  66528.000000  66528.000000   66528.000000   
mean       3.602715      3.605330      3.600133      3.602699    5135.847855   
std        0.417582      0.417625      0.417469      0.417573    9668.388846   
min        2.617000      2.621000      2.611000      2.617000       0.100000   
25%        3.270000      3.272000      3.268000      3.270000     982.575000   
50%        3.600000      3.601000      3.598500      3.600000    2380.650000   
75%        3.966000      3.969000      3.964000      3.966000    5570.225000   
max        4.487000      4.487000      4.444000      4.481000  407682.700000   

          Rec_count     Avg_price  
count  66528.000000  66528.000000  
mean      13.411827      3.602738  
std       11.505622      0.417538  
min        1.000000      2.618512  
25%        6.000000      3.269678  
50%       10.000000      3.599970  
75%       17.000000      3.966378  
max      119.000000      4.467341

<p>
This shows the statistical summary of all numeric-typed (int, float) columns.<br>

For example, the attribute "open" has 66528 counts, the mean value of this column is 3.602715, the standard deviation is 0.417582, the minimum value is 2.617000, 25th percentile is 3.272000, 50th percentile is 3.600000, 75th percentile is 3.966000, and the maximum value is 4.487000. <br>

However, what if we would also like to check all the columns including those that are of type object? <br><br>

You can add an argument <code>include = np.datetime64</code> inside the bracket. Let's try it again.

</p>


In [12]:
df.describe(include=np.datetime64)

C:\Users\melni\AppData\Local\Temp\ipykernel_5396\1761068203.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  df.describe(include=np.datetime64)


Ts
count                 66528
unique                66528
top     2022-11-11 14:38:00
freq                      1
first   2022-11-11 14:38:00
last    2022-12-29 23:59:00

<p>
Now it provides the statistical summary of all the columns, including object-typed attributes.<br>

We can now see how many unique values there, which one is the top value and the frequency of top value in the object-typed columns.<br>

Some values in the table above show as "NaN". This is because those numbers are not available regarding a particular column type.<br>

</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Question #3: </h1>

<p>
You can select the columns of a dataframe by indicating the name of each column. For example, you can select the three columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Where "column" is the name of the column, you can apply the method  ".describe()" to get the statistics of those columns as follows:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>

Apply the  method to ".describe()" to the columns 'low' and 'high'.

</div>


In [ ]:
# Write your code below and press Shift+Enter to execute 


<details><summary>Click here for the solution</summary>

```python
df[['Low', 'High']].describe()
```

</details>


<h2>Info</h2>
Another method you can use to check your dataset is:


It provides a concise summary of your DataFrame.

This method prints information about a DataFrame including the index dtype and columns, non-null values and memory usage.


In [13]:
# look at the info of "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 66528 entries, 0 to 66527
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Ts         66528 non-null  datetime64[ns]
 1   Open       66528 non-null  float64       
 2   High       66528 non-null  float64       
 3   Low        66528 non-null  float64       
 4   Close      66528 non-null  float64       
 5   Volume     66528 non-null  float64       
 6   Rec_count  66528 non-null  int64         
 7   Avg_price  66528 non-null  float64       
dtypes: datetime64[ns](1), float64(6), int64(1)
memory usage: 4.6 MB


<h1>Excellent! You have just completed the  Introduction Notebook!</h1>


### Thank you for completing this lab!

## Author

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Joseph Santarcangelo</a>

### Other Contributors

<a href="https://www.linkedin.com/in/mahdi-noorian-58219234/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Mahdi Noorian PhD</a>

Bahare Talayian

Eric Xiao

Steven Dong

Parizad

Hima Vasudevan

<a href="https://www.linkedin.com/in/fiorellawever/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ENSkillsNetwork20235326-2021-01-01" target="_blank">Fiorella Wenver</a>

<a href="https:// https://www.linkedin.com/in/yi-leng-yao-84451275/ " target="_blank" >Yi Yao</a>.

## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description                       |
| ----------------- | ------- | ---------- | ---------------------------------------- |
| 2022-08-23        | 2.4     | Malika     | Import micropip added and parameter for ipwidgets and tqdm |
| 2020-10-30        | 2.3     | Lakshmi    | Changed URL of the csv                   |
| 2020-09-22        | 2.2     | Nayef      | Added replace() method to remove '?'     |
| 2020-09-09        | 2.1     | Lakshmi    | Made changes in info method of dataframe |
| 2020-08-27        | 2.0     | Lavanya    | Moved lab to course repo in GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
